ipa-core/src/protocol/ipa_prf/shuffle/mod.rs

```diff
@@ -0,0 +1,167 @@
+    TS: SharedValue + CustomArray<Element = Boolean> + Field,
+    for<'a> &'a AdditiveShare<TS>: IntoIterator<Item = AdditiveShare<Boolean>>,
+    for<'a> &'a AdditiveShare<TV>: IntoIterator<Item = AdditiveShare<Boolean>>,
+    for<'a> &'a AdditiveShare<BK>: IntoIterator<Item = AdditiveShare<Boolean>>,
+    for<'a> <&'a AdditiveShare<TV> as IntoIterator>::IntoIter: Send,
+    for<'a> <&'a AdditiveShare<TS> as IntoIterator>::IntoIter: Send,
+{
+    let shuffle_input: Vec<AdditiveShare<BA112>> = input
+        .into_iter()
+        .map(|item| oprfreport_to_shuffle_input::<BA112, BK, TV, TS>(&item))
+        .collect::<Vec<_>>();
+
+    let shuffled = shuffle(ctx, shuffle_input).await?;
+
+    Ok(shuffled
+        .into_iter()
+        .map(|item| shuffled_to_oprfreport(&item))
+        .collect::<Vec<_>>())
+}
+
+// This function converts OprfReport to an AdditiveShare needed for shuffle protocol
+pub fn oprfreport_to_shuffle_input<YS, BK, TV, TS>(
+    input: &OprfReport<BK, TV, TS>,
+) -> AdditiveShare<YS>
+where
+    YS: CustomArray<Element = <BA112 as CustomArray>::Element> + SharedValue,
+    BK: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    TV: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    TS: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+{
+    let (mut y_left, mut y_right) = (YS::ZERO, YS::ZERO);
+    expand_array_in_place(&mut y_left, input.match_key.left(), 0);
+    expand_array_in_place(&mut y_right, input.match_key.right(), 0);
+
+    let mut offset = BA64::BITS as usize;
+
+    y_left.set(offset, input.is_trigger.left());
+    y_right.set(offset, input.is_trigger.right());
+
+    offset += 1;
+
+    expand_array_in_place(&mut y_left, input.breakdown_key.left(), offset);
+    expand_array_in_place(&mut y_right, input.breakdown_key.right(), offset);
+
+    offset += BK::BITS as usize;
+    expand_array_in_place(&mut y_left, input.trigger_value.left(), offset);
+    expand_array_in_place(&mut y_right, input.trigger_value.right(), offset);
+
+    offset += TV::BITS as usize;
+    expand_array_in_place(&mut y_left, input.timestamp.left(), offset);
+    expand_array_in_place(&mut y_right, input.timestamp.right(), offset);
+
+    AdditiveShare::<YS>::new(y_left, y_right)
+}
+
+// This function converts AdditiveShare obtained from shuffle protocol to OprfReport
+pub fn shuffled_to_oprfreport<YS, BK, TV, TS>(input: &AdditiveShare<YS>) -> OprfReport<BK, TV, TS>
+where
+    YS: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    BK: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    TV: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    TS: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+{
+    let match_key = extract_from_shared_array::<YS, BA64>(input, 0);
+
+    let mut offset = BA64::BITS as usize;
+
+    let is_trigger = AdditiveShare::<Boolean>::new(
```

Comment:
i can create a BA1 to encapsulate is_trigger so that all code looks uniform. But it would need refactoring is_trigger in more places. lmk if that looks better

Reply:
Boolean seems to me the right choice here. Using BA1 just adds another layer around it that probably makes it harder to use and doesn't seem to offer any advantages (beyond that you could have a generic concatenate functionality that can concatenate any BAx and BAy to a BAx+y)

Reply:
thats true but it would help keep everything in same format as commented above